local Export = {}

local Conversions = {
	Services = {
		["ReplicatedStorage"] = [[game:GetService("ReplicatedStorage")]],
		["ReplicatedFirst"] = [[game:GetService("ReplicatedFirst")]],
		["Workspace"] = [[game:GetService("Workspace")]],
	},
	Types = {
		["Vector3"] = function(Vec3)
			return `Vector3.new({Vec3.X}, {Vec3.Y}, {Vec3.Z})`
		end,
		["Vector2"] = function(Vec2)
			return `Vector2.new({Vec2.X}, {Vec2.Y})`
		end,
		["string"] = function(String)
			return `"{String}"`
		end,
		["number"] = function(Number)
			return `{tostring(Number)}`
		end,
		["Instance"] = function(Instance)
			return `game.{Instance:GetFullName()}`
		end,
		["userdata"] = function()
			return `newproxy(true)`
		end,
		["CFrame"] = function(CFrame: CFrame)
			local Position = CFrame.Position
			local Round = math.floor
			local Convert = function(n)
				return Round(math.deg(n))
			end

			local ax, ay, az = CFrame:ToEulerAnglesXYZ()
			return `CFrame.new({Round(Position.X)}, {Round(Position.Y)}, {Round(Position.Y)}) * CFrame.Angles(math.rad({Convert(
				ax
			)}), math.rad({Convert(ay)}), math.rad({Convert(az)})`
		end,
		["buffer"] = function(Buffer)
			return `buffer.create({buffer.len(Buffer)})`
		end,
	},
}

Conversions.Types["table"] = function(Table)
	local Converted = {}

	for Index, Value in Table do
		local VType = typeof(Value)
		local VIndex = typeof(Index)

		local ConvertValue = Conversions.Types[VType]
		if ConvertValue == nil then
			warn("Failed to convert type", VType, "(value)")
			continue
		end

		local ConvertIndex = Conversions.Types[VIndex]
		if ConvertIndex == nil then
			warn("Failed to convert type", VIndex, "(index)")
			continue
		end

		local ValueString = ConvertValue(Value)
		local IndexString = ConvertIndex(Index)

		table.insert(Converted, `[{IndexString}] = {ValueString}`)
	end
	return table.concat(Converted, ", ")
end

local RealServices = {}
for ServiceName, ServiceGet in Conversions.Services do
	local Service = game:GetService(ServiceName)
	RealServices[Service.Name] = ServiceGet
end
Conversions.Services = RealServices

function Export:Generate(Event, CallId)
	local Code = "-- Failed to generate"

	local CallData = Event.Calls[CallId]
	if not CallData then
		return Code
	end

	local Instance = Event.Instance
	if not Instance then
		return Code
	end

	local FullName = Instance:GetFullName()
	for Original, New in Conversions.Services do
		local Result = FullName:sub(1, #Original)
		if Result == Original then
			FullName = `{New}["{FullName:gsub(Original .. ".", "")}"]`
		end
	end

	local Args = {}
	for Index, Value in CallData.Args do
		local Type = typeof(Value)
		local Conversion = Conversions.Types[Type]
		if Conversion == nil then
			print("Type", type, "not recognized")
			continue
		end
		table.insert(Args, Conversion(Value))
	end

	if CallData.Nils > 0 then
		for i = 1, CallData.Nils do
			table.insert(Args, "nil")
		end
	end

	local StrArgs = table.concat(Args, ", ")

	Code = `-- Generated by Obsidian\n{FullName}:{CallData.Caller.Method}({StrArgs})`

	return Code
end

return Export
